src/theory/theory_engine_proof_generator.h

```diff
@@ -0,0 +1,69 @@
+/*********************                                                        */
+/*! \file theory_engine.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **  Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2019 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief The theory engine proof generator
+ **/
+
+#include "cvc4_private.h"
+
+#ifndef CVC4__THEORY_ENGINE_PROOF_GENERATOR_H
+#define CVC4__THEORY_ENGINE_PROOF_GENERATOR_H
+
+#include <memory>
+
+#include "context/cdhashmap.h"
+#include "context/context.h"
+#include "expr/lazy_proof.h"
+#include "expr/proof_generator.h"
+#include "expr/proof_node_manager.h"
+#include "theory/trust_node.h"
+
+namespace CVC4 {
+
+/**
+ * A simple proof generator class used by the theory engine. This class
+ * stores proofs for TheoryEngine::getExplanation.
+ *
+ * Notice that this class could be made general purpose. Its main feature is
+ * storing lazy proofs for facts in a context-dependent manner.
+ */
+class TheoryEngineProofGenerator : public ProofGenerator
+{
+  typedef context::
+      CDHashMap<Node, std::shared_ptr<LazyCDProof>, NodeHashFunction>
+          NodeLazyCDProofMap;
+
+ public:
+  TheoryEngineProofGenerator(ProofNodeManager* pnm, context::UserContext* u);
+  ~TheoryEngineProofGenerator() {}
+  /**
+   * Make trust explanation. Called when lpf has of proof of (=> exp lit). This
+   * stores lpf in the map d_proofs below and returns the trust node for this
+   * propagation, which has TrustNodeKind TrustNodeKind::PROP_EXP.
+   */
```

Comment:
Also mention that if `(=> exp lit)` is already registered the new proof is ignored (and say why that's OK).